wgengine/wgcfg/parser.go

```diff
@@ -122,62 +117,70 @@ func FromUAPI(r io.Reader) (*Config, error) {
 	return cfg, nil
 }
 
-func (cfg *Config) handleDeviceLine(key, value string) error {
-	switch key {
-	case "private_key":
-		k, err := parseKeyHex(value)
-		if err != nil {
+func parseKeyHex(s []byte, dst []byte) error {
+	n, err := hex.Decode(dst, s)
+	if err != nil {
+		return &ParseError{"Invalid key: " + err.Error(), string(s)}
+	}
+	if n != wgkey.Size {
+		return &ParseError{"Keys must decode to exactly 32 bytes", string(s)}
+	}
+	return nil
+}
+
+func (cfg *Config) handleDeviceLine(key, value mem.RO, valueBytes []byte) error {
+	if key.EqualString("private_key") {
+		// wireguard-go guarantees not to send zero value; private keys are already clamped.
+		if err := parseKeyHex(valueBytes, cfg.PrivateKey[:]); err != nil {
 			return err
 		}
-		// wireguard-go guarantees not to send zero value; private keys are already clamped.
-		cfg.PrivateKey = wgkey.Private(*k)
-	case "listen_port":
+	} else if key.EqualString("listen_port") || key.EqualString("fwmark") {
 		// ignore
-	case "fwmark":
-		// ignore
-	default:
+	} else {
 		return fmt.Errorf("unexpected IpcGetOperation key: %v", key)
 	}
 	return nil
 }
 
-func (cfg *Config) handlePublicKeyLine(value string) (*Peer, error) {
-	k, err := parseKeyHex(value)
-	if err != nil {
+func (cfg *Config) handlePublicKeyLine(value mem.RO, valueBytes []byte) (*Peer, error) {
+	p := Peer{}
+	if err := parseKeyHex(valueBytes, p.PublicKey[:]); err != nil {
 		return nil, err
 	}
-	cfg.Peers = append(cfg.Peers, Peer{})
-	peer := &cfg.Peers[len(cfg.Peers)-1]
-	peer.PublicKey = *k
-	return peer, nil
+	cfg.Peers = append(cfg.Peers, p)
+	return &cfg.Peers[len(cfg.Peers)-1], nil
 }
 
-func (cfg *Config) handlePeerLine(peer *Peer, key, value string, valueBytes []byte) error {
-	switch key {
-	case "endpoint":
-		err := json.Unmarshal(valueBytes, &peer.Endpoints)
-		if err != nil {
+func (cfg *Config) handlePeerLine(peer *Peer, key, value mem.RO, valueBytes []byte) error {
+	if key.EqualString("endpoint") {
+		if err := json.NewDecoder(mem.NewReader(value)).Decode(&peer.Endpoints); err != nil {
```

Comment:
I think making that NewReader is worse than just using valueBytes like before

Reply:
I experimented with both and didn't notice too much of a difference, will switch to using valueBytes

Reply:
Because your benchmark never exercises this line of code.

Your input string is:

```
private_key=a87242cf17a8acc459b444e1fac8bb5e848784b97e41adf3ad00d34f43327350
public_key=f7b7f5c3aa2b5a9ec7363dc1753d5872fae63c4ec9da3591c56334d64b2d7a1b
protocol_version=1
replace_allowed_ips=true
allowed_ip=10.0.0.1/32
public_key=f7b7f5c3aa2b5a9ec7363dc1753d5872fae63c4ec9da3591c56334d64b2d7a1b
protocol_version=1
replace_allowed_ips=true
allowed_ip=10.0.0.1/32
public_key=f7b7f5c3aa2b5a9ec7363dc1753d5872fae63c4ec9da3591c56334d64b2d7a1b
protocol_version=1
replace_allowed_ips=true
allowed_ip=10.0.0.1/32
public_key=f7b7f5c3aa2b5a9ec7363dc1753d5872fae63c4ec9da3591c56334d64b2d7a1b
protocol_version=1
replace_allowed_ips=true
allowed_ip=10.0.0.1/32
```

It doesn't contain `endpoint`.

Reply:
Added and now the benchmarks reflect that